Incorrect progressbar alert 10899

adam-james-v opened the pull request:
The goal-met predicate is tested directly with mock data to make the tests more readable.

The change to `metabase.pulse/goal-met?` fixes #10899 while preserving the goal-met behaviour for any non-progress
goal. The behaviour should now be as follows:

Edit: After some discussion, it's clear that Progress Bar and Time Series comparators _should be the same_. The Table is edited to reflect this. It means the change is simpler than originally thought, yay!

```
| alert_above? | goal | val | goal-met? |
+--------------+------+-----+-----------+
|      t       |  5   |  4  |     f     |
|      t       |  5   |  5  |     t     |
|      t       |  5   |  6  |     t     |
|      f       |  5   |  4  |     t     |
|      f       |  5   |  5  |     f     | <--- this is new behaviour
|      f       |  5   |  6  |     f     | 
```

adam-james-v committed:
Comparison for goals depends on alert_above_goal and if progressbar.

This commit sets up the timeseries predicate and writes tests for what I think the correct behaviours should be:

timeseries, alert above -> (<= goal-value value)
 - (<= 5.9 6) true  -> alert sent
 - (<= 6 6)   true  -> alert sent
 - (<= 6.1 6) false -> alert not sent

progressbar, alert above -> (<= goal-value value)
 - (<= 5.9 6) true  -> alert sent
 - (<= 6 6)   true  -> alert sent
 - (<= 6.1 6) false -> alert not sent

timeseries, alert below -> (>= goal-value value)
 - (>= 6.1 6) true   -> alert sent
 - (>= 6 6)   true   -> alert sent. NOTE: this is to match prior behaviour, but I don't understand why it is this way?
 - (>= 5.9 6) faluse -> alert not sent

progressbar, alert below -> (> goal-value value)
 - (> 6.1 6) true   -> alert sent
 - (> 6 6)   false  -> alert not sent. NOTE: this is what should fix the bug (#10899)
 - (> 5.9 6) faluse -> alert not sent

I think there may be a cleaner way to write and present the tests, so the next commit(s) will address that.

adam-james-v committed:
Simplified Tests for goal-met? predicate.

The goal-met predicate is tested directly with mock data to make the tests more readable.

The change to `metabase.pulse/goal-met?` fixes #10899 while preserving the goal-met behaviour for any non-progress
goal. The behaviour should now be as follows:

```
| Timeseries? | alert_above? | goal | val | goal-met? |
+-------------+--------------+------+-----+-----------+
|     f       |      t       |  5   |  4  |     f     |
|     f       |      t       |  5   |  5  |     t     |
|     f       |      t       |  5   |  6  |     t     |
|     f       |      f       |  5   |  4  |     t     |
|     f       |      f       |  5   |  5  |     f     | <--- this is new behaviour
|     f       |      f       |  5   |  6  |     f     |

|     t       |      t       |  5   |  4  |     f     |
|     t       |      t       |  5   |  5  |     t     |
|     t       |      t       |  5   |  6  |     t     |
|     t       |      f       |  5   |  4  |     t     |
|     t       |      f       |  5   |  5  |     t     |
|     t       |      f       |  5   |  6  |     f     |
```

Otherwise, alerts should be triggered as before.

adam-james-v committed:
added issue number to test

adam-james-v committed:
fixed alignment issue

adam-james-v commented:
@noah, there was a prior PR for #10899 where you mentioned subtle behaviour differences with progress bar goals and other goals [here](https://github.com/metabase/metabase/pull/18058#issuecomment-939038983)

This PR adjusts the comparator used when there's a progress bar goal whose value is equal to the goal. It used to be >=, but is now >.

I'm just curious if I've even thought through this properly? Because it also feels incorrect to me that in the other cases we have both <= and >=. This feels like an incorrect handling of the equals case?

adam-james-v committed:
swapped goal and value around in comparison for readability

noahmoss commented:
Also, if you rebase or merge in the lastest `master`, it should fix the CI failures you're seeing

adam-james-v committed:
Fixing nits

adam-james-v committed:
Timeseries and progress bar now follow same logic

I asked for clarification about when alerts fire in #ama-product in Metabase's Slack. There was consensus around the
idea that 'reaches the goal' does generally imply 'meets the goal', so we can use the same comparators between
progress bar and timeseries goals now.

This simplifies the goal-met? function again because it's no longer necessary to check if the given data is a
timeseries goal.

adam-james-v committed:
Adjusted text in goal line alert modal

adam-james-v committed:
ran prettier
